storage/storage.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2017 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// Adapted from istio/proxy/test/backend/echo with error handling and
+// concurrency fixes and making it as low overhead as possible
+// (no std output by default)
+
+package storage // import "istio.io/fortio/storage"
+
+import (
+	"time"
+
+	"istio.io/fortio/log"
+
+	"cloud.google.com/go/datastore"
+	"golang.org/x/net/context"
+)
+
+// DatastoreClient is a struct for interacting with Google Cloud Datastore
+type DatastoreClient struct {
+	client *datastore.Client
+}
+
+// IstioPerfResult is the data structure for uploading a test run result to Google Cloud Datastore
+type IstioPerfResult struct {
```

Comment:
Can we use the (Http)RunnerResults directly ?

Reply:
I'd expect what we upload to be more or less different from what those internal data structures. The ResultJSON here should probably just be the serialization of RunnerResults. And we only put fields we explicitly want to index here.